### DISCUSSION TITLE
A case for re-instating the 'edit' button

### DISCUSSION THREAD

**Discussion by willblackmore** [outline/outline#3311] [Ideas]
Since a recent update where we lost the 'Edit' button and all documents are always in editing mode I feel the UX has taken quite a significant step backwards in 3 important ways.

**You can no longer click links**
Once a link has been created, to follow the link you have to click the text, then click the little icon (with no label) on the tooltip which appears above the text.

Not all our users are technical: _I had to put a message out on our all-agency Slack channel explaining how to click links._ 🙈
![links](https://user-images.githubusercontent.com/1260445/161035804-f2f64e12-3382-4ca5-b1c7-15f8d43969aa.gif)

**Accidental edits**
We have the Outline Slack app set up to post updates in a specific Slack channel, and I see constant accidental edits. It happens to me too. Whole chunks of documents have gone missing simply because someone selected some text to copy and accidentally hit a key without realising.

**Slack app spam**
As I mentioned above, we have the Slack app installed. Now, while editing a document we get several "[name] has updated a document" messages as we go rather than one update once we're done.
<img width="686" alt="Screenshot 2022-03-31 at 11 37 37" src="https://user-images.githubusercontent.com/1260445/161036456-d66ad263-0657-43bc-a50f-81ada2421b31.png">

Does anyone else agree?

**Comment by Limezy** [comment]
What about having a "lock" button ? That would at least prevent the accidental edits.
For the slack updates, I don't use it but I understand that it can be quite cumbersome.
For the UX "broken" about clicking links, It's true that it could be optimized, even if I have now taken the habit to click "outside" of the editor every time I need to use links.

**Comment by roovyshapiro** [comment]
One of the features I loved the most was not having an edit button. I actually summitted a request to remove the draft/publish feature for new documents as well.
I personally feel that a documentation platform will only be implemented successfully if it removes as much barriers to contributions as possible. We work in a fast paced environment and it sounds silly, but having to wait and click edit and then publish is enough of a barrier where it prevents a lot of users from feeling comfortable to make a quick contribution or copy/paste into a document before they forget. 
So while I agree that links are a bit frustrating, I think the editing "mistakes" are worth it.

To fix the link issue, I think Outline can benefit from the way that Notion has implemented their various Links.

1. A link to a page is a standalone full width "block" so clicking anywhere on the line opens it and the "block" has to be edited to change it.
2. Clicking on a hyperlink of specific text always opens it. To edit it, you have to put your cursor in the text, or hover over the link to edit it. So opening is it the default and editing requires an extra step.
3. Links to URLS are their own "block" which again have clicking on them open it by default.

Here's a GIF of me showing how this process works: 
https://i.imgur.com/v9iDvFl.gif

Notion also uses always live page edits, but fixing the way links are accessed will hopefully solve most of this, and then the "Accidental edits" are worth the frustration IMO.

**Reply by willblackmore** [reply]
I understand the ethos behind the decision, but the fact that most of our employees are struggling with it suggests that there are better ways to introduce this feature.

If an actual edit button isn't an option, how about using a double click to enter edit mode? It's a solid convention used across many different apps and operating systems.

**Double click** any content on a page to edit. Hit **Esc** to finish editing.

While it's still a barrier, I think it's a pretty small one.

Whatever the method, I'd also recommend some kind of clear visual indication that the user is editing.

Right now, when a page loads you can actually click links normally. Click any text (except links) and you subtly enter edit mode and lose that ability. Click the sidebar and you leave edit mode (again, very subtly).
![example](https://user-images.githubusercontent.com/1260445/161932336-d7130c63-ccd8-40a0-a054-6c5f6b06dacf.gif)

The only visual indication you are in editing mode is that the heading level shows next to each heading (h1, h2, h3 etc.).

**Reply by roovyshapiro** [reply]
_"Whatever the method, I'd also recommend some kind of clear visual indication that the user is editing."_

That I can agree with at least. Maybe a colored bar at the top. Or maybe an obvious icon or text that says "editing...." and turns into "saving..." or "saved"

But I really would not like a double click/ESC or some kind of editing/saving mechanism in place.

I think that with regards to links I would prefer a solution which has clicking links always default to opening them in a new tab even while in editing mode. 
To edit the link, you'd have to either 
- hover over it, wait for a small popup with the option to edit and then select edit
- double click to edit it instead of single clicking
- right click to edit it
I think that would be a better solution to this problem.

**Reply by willblackmore** [reply]
> But I really would not like a double click/ESC or some kind of editing/saving mechanism in place.

No, honestly, neither do I - I'd rather have a visual toggle which switches a read-only state on and off. I feel like this could be a preference set at the user level (though as an admin I'd like to be able to set the default).

I do like your first suggestion of the links always being clickable but a UI which shows on hover allowing you to edit. And also selecting the text and using CMD/CTRL + K should still work for those who like keyboard shortcuts.

**Reply by mweth** [reply]
How about a "confirm edit" button to at least let us discard inadvertent changes.  This is a barrier for use.

**Comment by edwarddarrah** [comment]
I agree that it is problematic for the page to be in live edit mode at all times. An alternate solution could be to require a page that has been edited to be Re-Published. Subsequently, any edit to an existing page would make a draft; the user would need to select Re-Published for the edits to go live on the page. 

Or, for existing pages, there can be a toggle button where Publish would be if it were a new page. The toggle would be set to "Read" when entering an existing page. To edit a page, the user needs to toggle to "Edit" to allow the page to be edited. Then, the user would need to toggle back to Read mode for the edits to be updated on the page. The platform [Slab](https://slab.com/) handles edits like this. 

![CleanShot 2022-11-20 at 03 13 51](https://user-images.githubusercontent.com/96745845/202892374-f781bf3d-179f-4082-b5a3-07b715b533b5.png)

**Comment by roovyshapiro** [comment]
Perhaps instead, the slider could simply display what mode it's currently
in. I don't want to implement a manual step to publish changes to a
document. I think live editing is great. But it may be helpful to have a
visual indicator like the slider to let the user know which mode they're in
while still maintaining the mode change as it exists today.

**Reply by edwarddarrah** [reply]
@roovyshapiro I think this is a viable solution. The more I have given it thought; I also enjoy live editing as well. Your suggestions do a good job of covering both preferences.

**Comment by ChetanGoti** [comment]
There is now preference for disabling the seamless editing feature, so documents are opened in Read only mode and can be edited.

See https://github.com/outline/outline/issues/3132

**Reply by roovyshapiro** [reply]
I think this is a decent solution. I wish this setting could be enabled per collection though
